-- should be updated when roblox adds new global | @jLn0n

return {
    -- Luau Functions
    ["assert"] = "function",
    ["error"] = "function",
    ["getfenv"] = "function",
    ["getmetatable"] = "function",
    ["ipairs"] = "function",
    ["loadstring"] = "function",
    ["newproxy"] = "function",
    ["next"] = "function",
    ["pairs"] = "function",
    ["pcall"] = "function",
    ["print"] = "function",
    ["rawequal"] = "function",
    ["rawget"] = "function",
    ["rawlen"] = "function",
    ["rawset"] = "function",
    ["select"] = "function",
    ["setfenv"] = "function",
    ["setmetatable"] = "function",
    ["tonumber"] = "function",
    ["tostring"] = "function",
    ["unpack"] = "function",
    ["xpcall"] = "function",
    ["type"] = "function",
    ["typeof"] = "function",

    -- Luau Functions (Deprecated)
    ["collectgarbage"] = "function",

    -- Luau Variables
    ["_G"] = "table",
    ["_VERSION"] = "string",

    -- Luau Tables
    ["bit32"] = "table",
    ["coroutine"] = "table",
    ["debug"] = "table",
    ["math"] = "table",
    ["os"] = "table",
    ["string"] = "table",
    ["table"] = "table",
    ["utf8"] = "table",
    ["buffer"] = "table",

    -- Roblox Functions
    ["DebuggerManager"] = "function",
    ["delay"] = "function",
    ["gcinfo"] = "function",
    ["PluginManager"] = "function",
    ["require"] = "function",
    ["settings"] = "function",
    ["spawn"] = "function",
    ["tick"] = "function",
    ["time"] = "function",
    ["UserSettings"] = "function",
    ["wait"] = "function",
    ["warn"] = "function",

    -- Roblox Functions (Deprecated)
    ["Delay"] = "function",
    ["ElapsedTime"] = "function",
    ["elapsedTime"] = "function",
    ["printidentity"] = "function",
    ["Spawn"] = "function",
    ["Stats"] = "function",
    ["stats"] = "function",
    ["Version"] = "function",
    ["version"] = "function",
    ["Wait"] = "function",
    ["ypcall"] = "function",

    -- Roblox Variables
    ["game"] = "Instance",
    ["plugin"] = "Instance",
    ["shared"] = "table",
    ["workspace"] = "Instance",

    -- Roblox Variables (Deprecated)
    ["Game"] = "Instance",
    ["Workspace"] = "Instance",

    -- Roblox Tables
    ["Axes"] = "table",
    ["BrickColor"] = "table",
    ["CatalogSearchParams"] = "table",
    ["CFrame"] = "table",
    ["Color3"] = "table",
    ["ColorSequence"] = "table",
    ["ColorSequenceKeypoint"] = "table",
    ["DateTime"] = "table",
    ["DockWidgetPluginGuiInfo"] = "table",
    ["Enum"] = "Enums",
    ["Faces"] = "table",
    ["FloatCurveKey"] = "table",
    ["Font"] = "table",
    ["Instance"] = "table",
    ["NumberRange"] = "table",
    ["NumberSequence"] = "table",
    ["NumberSequenceKeypoint"] = "table",
    ["OverlapParams"] = "table",
    ["PathWaypoint"] = "table",
    ["PhysicalProperties"] = "table",
    ["Random"] = "table",
    ["Ray"] = "table",
    ["RaycastParams"] = "table",
    ["Rect"] = "table",
    ["Region3"] = "table",
    ["Region3int16"] = "table",
    ["RotationCurveKey"] = "table",
    ["SharedTable"] = "table",
    ["task"] = "table",
    ["TweenInfo"] = "table",
    ["UDim"] = "table",
    ["UDim2"] = "table",
    ["Vector2"] = "table",
    ["Vector2int16"] = "table",
    ["Vector3"] = "table",
    ["Vector3int16"] = "table",
}